### PR TITLE
[Week4] - seungwoo

### PR DESCRIPTION
**Pull request by Gonzo-Dev-lab**

### CODE REVIEW

**Review by kimhamyong — APPROVED** [review]
앞으로는 week4-seungwoo 파일만 업로드 해주세요!

File: week4-seungwoo/task1.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    model = tf.keras.models.Sequential([
+    tf.keras.layers.Dense(256),
+    tf.keras.layers.BatchNormalization(),
+    tf.keras.layers.ReLU(),
+    tf.keras.layers.Dropout(0.3),
+    tf.keras.layers.Dense(128),
+    tf.keras.layers.ReLU(),
+    tf.keras.layers.BatchNormalization(),
+    tf.keras.layers.Dropout(0.3),
+    tf.keras.layers.Dense(10, activation='sigmoid'),
```
Comment (kimhamyong) [inline week4-seungwoo/task1.py:68]: tf.keras.layers.Dense(10, activation='sigmoid')는 다중 클래스 분류 문제에서는 적합하지 않습니다. 대신 Dense(10, activation='softmax')를 사용하는 것이 올바릅니다. 이유는 다음과 같습니다:

- sigmoid 활성화 함수:
각 출력 노드에서 독립적으로 확률 값을 반환.
출력값은 [0, 1] 범위이며, 모든 출력값의 합이 1이 아님.
이 함수는 다중 레이블 분류 문제에 적합 (예: 한 샘플이 여러 클래스에 속할 수 있는 경우).
- softmax 활성화 함수:
모든 출력값의 합이 1이 되도록 확률을 정규화.
각 클래스에 대한 확률 값을 출력하며, 출력값이 서로 종속적.
이 함수는 다중 클래스 분류 문제에 적합 (예: 한 샘플이 정확히 하나의 클래스에 속할 때).

- MNIST 데이터셋은 **10개의 클래스(0~9 숫자)**로 구성된 다중 클래스 분류 문제입니다.
- 각 이미지가 정확히 하나의 클래스에 속하므로, 출력층에서 softmax 활성화를 사용해야 합니다.
- sigmoid는 모든 클래스가 독립적인 확률로 출력되기 때문에, 다중 클래스 분류 문제에서는 부적합합니다.

```python
Dense(10, activation='softmax')
```

File: week4-seungwoo/task1.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# Adam 옵티마이저, sparse categorical crossentropy 손실 함수, accuracy 메트릭으로 모델을 컴파일
+model = create_model()
+model.compile(
+    optimizer=Adam(),
+    loss='sparse_categorical_crossentropy',
+    metrics=['accuracy']
+)
+
+# 배치 크기를 64로, 에포크 수를 5로 설정하여 모델을 훈련
+model.fit(x_train, y_train, batch_size=64, epochs=5)
```
Comment (kimhamyong) [inline week4-seungwoo/task1.py:91]: - validation_split 옵션을 활용해 훈련 데이터의 일부를 검증 데이터로 설정하면 학습 및 검증 성능을 동시에 확인할 수 있습니다.

- 추가적으로, 학습 도중 검증 손실이 더 이상 감소하지 않을 때 학습을 조기 종료하도록 EarlyStopping 콜백을 추가하면 불필요한 에포크 실행을 방지할 수 있습니다. (필수 사항 X)

```python
from tensorflow.keras.callbacks import EarlyStopping
early_stopping = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)
history = model.fit(x_train, y_train, batch_size=64, epochs=50, validation_split=0.2, callbacks=[early_stopping])
```

Comment (kimhamyong) [inline week4-seungwoo/task1.py:91]: 데이터셋을 수동으로 섞는 대신 TensorFlow의 tf.data API를 활용하면 더 효율적이고 가독성이 높은 코드를 작성할 수 있습니다.

```python
train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
train_dataset = train_dataset.shuffle(buffer_size=10000).batch(64)

model.fit(train_dataset, epochs=5)
```

**Review by kimhamyong — APPROVED** [review]
Task1: 잘 수행하셨습니다.

**Review by giyomdabin — APPROVED** [review]
task2: 주어진 과제에 대해 잘 수행하셨습니다. 수고하셨습니다!
